microhapulator/workflows/preproc-single.smk

```diff
@@ -37,22 +38,26 @@ rule fastqc:
         symlink(outfile.name, linkfile)
 
 
-rule fastq_reads:
+rule filter_ambiguous:
     input:
         lambda wildcards: sorted([fq for fq in config["readfiles"] if wildcards.sample in fq]),
     output:
-        fastq="analysis/{sample}/reads.fastq",
+        filtered="analysis/{sample}/{sample}-ambig-filtered.fastq",
+        copied_fq="analysis/{sample}/reads.fastq",
+        counts="analysis/{sample}/{sample}-ambig-read-counts.txt",
+    params:
+        ambig_thresh=config["ambiguous_thresh"],
     run:
         assert len(input) == 1
-        if input[0].endswith(".gz"):
-            shell("gunzip -c {input[0]} > {output.fastq}")
-        else:
-            shell("cp {input[0]} {output.fastq}")
+        out_prefix = f"analysis/{wildcards.sample}/{wildcards.sample}"
+        ambig_filter = AmbigSingleReadFilter(input[0], out_prefix, params.ambig_thresh)
+        ambig_filter.filter()
+        shell("cp {output.filtered} {output.copied_fq}")
```

Comment:
I'm wondering if we  really need to do the copying step at all? The compression is now handled in the new `filter.py` module. And we have the original fastqs already linked/copied in `workdir/seqs`.  Snakemake will only ever need the filtered reads beyond this step.

Reply:
Many of our workflows use the `analysis/` and `seq/` folder layout in the working directory. By default, Fastq input files are symlinked to `seq/`, but for casework we often provide a way to copy instead of symlink so that full data provenance is maintained within the working directory. So this is more driven by consistency across workflows and casework provenance than it is by technical necessity.

Reply:
Ah, I see. I think maintaining copies or symlinks to the original reads in `seq/` is sufficient. Anything else we do in `analysis/` is fair game in terms of what we think is best for the workflow.

Reply:
Upon further consideration, I realized an advantage to creating `analysis/{sample}/reads.fastq` is that the input file name is the same for the `map_and_sort_index` rule for paired and singled end samples. Without this, we would have to handle the two cases separately since reads are mapped after merging for paired ended samples while they are mapped after filtering for single ended samples. I opted to keep the copying step in but changed the file name to `analysis/{sample}/preprocessed-reads.fq` to hopefully make it clear that these are not the original reads.